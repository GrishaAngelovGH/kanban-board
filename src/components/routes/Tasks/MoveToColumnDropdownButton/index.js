import MoveToColumnDropdownButton from "./MoveToColumnDropdownButton"

export default MoveToColumnDropdownButton