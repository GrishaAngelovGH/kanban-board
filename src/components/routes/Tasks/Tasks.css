.accordion {
  height: 80vh;
  overflow: auto;
}

.accordion-header > button {
  display: block;
}

.task-actions button {
  margin-right: 10px;
}

@media only screen and (max-width: 576px) {
  .dropdown-menu-header {
    width: 100%;
  }

  .task-actions button {
    width: 100%;
    margin-right: 0;
    margin-bottom: 5px;
  }
}
