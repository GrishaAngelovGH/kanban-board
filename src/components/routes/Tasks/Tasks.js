import { useState } from "react"
import { Link } from "react-router-dom"

import boardRepository from "persistent/persistentKanbanBoardRepository"
import userRepository from "persistent/persistentUserRepository"

import useBackgroundImage from "hooks/useBackgroundImage"

import Avatar from "components/Avatar"
import RichTextDescription from "components/RichTextDescription"
import Collaborators from "./Collaborators"
import MoveToColumnDropdownButton from "./MoveToColumnDropdownButton"

import Accordion from "react-bootstrap/Accordion"
import Badge from "react-bootstrap/Badge"
import Button from "react-bootstrap/Button"

import "./Tasks.css"

const Tasks = ({ userId, showToastWithMessage }) => {
  const [activeKey, setActiveKey] = useState("")

  const backgroundImage = useBackgroundImage()
  const user = userRepository.findUserById(userId)

  const relevantColumns = boardRepository.getColumnsWithAssignedTasksForUser(user.id)
  const allColumns = boardRepository.getColumns()

  return (
    <div className="row g-0 vh-100" style={{ backgroundImage: backgroundImage ? `url(${backgroundImage})` : "none", backgroundSize: "cover" }}>
      <div className="col-12">
        <div className="row g-0 bg-light p-2 shadow align-items-center">
          <div className="col-2 col-md-1">
            <Link to="/" className="btn btn-light bi bi-arrow-left border"></Link>
          </div>
          <div className="col-10 col-md-9 col-lg-10">
            <h3 className="m-0 text-center">Tasks assigned to {user.name}</h3>
            <p className="text-secondary text-center">Click on any item to expand/collapse its content</p>
          </div>
          <div className="col-1 col-md-2 col-lg-1 d-none d-md-block text-center">
            <Avatar user={user} size={80}></Avatar>
          </div>
        </div>

        <div className="row g-0 mt-3 justify-content-center">
          <div className="col-10">
            <Accordion activeKey={activeKey} onSelect={setActiveKey}>
              {
                relevantColumns.map((column, i) => (
                  <Accordion.Item key={column.id} eventKey={`${i}`}>
                    <Accordion.Header className={column.markedAsDone ? "done" : ""}>
                      <h3>
                        {column.title}
                        <Badge bg="primary ms-2">
                          {column.items.length}
                        </Badge>
                      </h3>
                      <p>{column.description}</p>
                    </Accordion.Header>
                    <Accordion.Body>
                      {
                        column.items.map(v => (
                          <div key={v.id} className="mt-3 border border-3 shadow rounded p-3 bg-light">
                            <h3>{v.title} {v.isLocked && <i className="bi bi-lock"></i>}</h3>

                            <RichTextDescription description={v.description} className="mb-3" />

                            {
                              v.priority.length > 0 && (
                                <p className="text-capitalize">Priority: {v.priority}</p>
                              )
                            }

                            <Collaborators ids={v.assignedIds} userId={userId} />

                            <div className="task-actions">
                              <MoveToColumnDropdownButton
                                columns={allColumns}
                                column={column}
                                taskId={v.id}
                                showToastWithMessage={showToastWithMessage}
                                setActiveKey={setActiveKey}
                                disabled={v.isLocked}
                              />

                              <Button
                                variant="danger"
                                size="sm"
                                onClick={() => {
                                  boardRepository.removeAssignedUserFromTask(v.id, column.id, userId)
                                  setActiveKey("")
                                  showToastWithMessage(`The user is successfully unassigned from "${v.title}" task`)
                                }}
                              >
                                Unassign Me
                              </Button>

                              <Button
                                variant="danger"
                                size="sm"
                                onClick={() => {
                                  v.assignedIds.forEach(id => {
                                    id !== userId && boardRepository.removeAssignedUserFromTask(v.id, column.id, id)
                                  })

                                  setActiveKey("")
                                  showToastWithMessage(`The remaining collaborators have been successfully unassigned from task "${v.title}"`)
                                }}
                              >
                                Leave only me
                              </Button>
                            </div>
                          </div>
                        ))
                      }
                    </Accordion.Body>
                  </Accordion.Item>
                ))
              }
            </Accordion>
          </div>
        </div>
      </div>
    </div>
  )
}

export default Tasks