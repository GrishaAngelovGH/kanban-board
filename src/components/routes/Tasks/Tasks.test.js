import { render } from "@testing-library/react"

import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom"

import Tasks from "./Tasks"

jest.mock('persistent/persistentUserRepository')

test("should render Tasks route component", () => {
  const view = render(
    <Router>
      <Routes>
        <Route path="/" element={<Tasks userId="1" />} />
      </Routes>
    </Router>
  )

  expect(view).toMatchSnapshot()
})