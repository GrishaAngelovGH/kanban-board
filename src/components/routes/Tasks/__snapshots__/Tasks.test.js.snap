// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Tasks route component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="row g-0 vh-100"
        style="background-image: url(geometric-triangle-shapes-background.jpg); background-size: cover;"
      >
        <div
          class="col-12"
        >
          <div
            class="row g-0 bg-light p-2 shadow align-items-center"
          >
            <div
              class="col-2 col-md-1"
            >
              <a
                class="btn btn-light bi bi-arrow-left border"
                href="/"
              />
            </div>
            <div
              class="col-10 col-md-9 col-lg-10"
            >
              <h3
                class="m-0 text-center"
              >
                Tasks assigned to 
                John Smith
              </h3>
              <p
                class="text-secondary text-center"
              >
                Click on any item to expand/collapse its content
              </p>
            </div>
            <div
              class="col-1 col-md-2 col-lg-1 d-none d-md-block text-center"
            >
              <img
                alt="user"
                class="rounded-circle m-1"
                src="image_1"
                width="80"
              />
            </div>
          </div>
          <div
            class="row g-0 mt-3 justify-content-center overflow-auto tasks"
          >
            <div
              class="col-10"
            >
              <p
                class="bg-white rounded text-center p-1 undefined"
              >
                Completed 
                0
                 task
                s
                 of 
                2
                 (
                0
                %)
              </p>
              <div
                class="accordion"
              >
                <div
                  class="accordion-item"
                >
                  <h2
                    class="accordion-header"
                  >
                    <button
                      aria-expanded="false"
                      class="accordion-button collapsed"
                      type="button"
                    >
                      <h3>
                        Backlog
                        <span
                          class="badge bg-primary ms-2"
                        >
                          2
                        </span>
                      </h3>
                      <p>
                        Prioritized list of features that should be implemented as part of a project.
                      </p>
                    </button>
                  </h2>
                  <div
                    class="accordion-collapse collapse"
                  >
                    <div
                      class="accordion-body"
                    >
                      <div
                        class="mt-3 border border-3 shadow rounded p-3 bg-light"
                      >
                        <h3>
                          Implement ConfirmButton component for ProductModal
                           
                        </h3>
                        <div
                          class="mb-3"
                        >
                          <span>
                            <span
                              class=""
                            >
                              A ConfirmButton component should be implemented to allow confirming selected product features in ProductModal.
                            </span>
                          </span>
                        </div>
                        <p
                          class="text-capitalize"
                        >
                          Priority: 
                          medium
                        </p>
                        <div />
                        <div
                          class="task-actions"
                        >
                          <div
                            class="dropdown-menu-header dropdown btn-group"
                            role="group"
                          >
                            <button
                              aria-expanded="false"
                              class="dropdown-toggle btn btn-primary btn-sm"
                              type="button"
                            >
                              Move to other column
                            </button>
                          </div>
                          <button
                            class="btn btn-danger btn-sm"
                            type="button"
                          >
                            Unassign Me
                          </button>
                          <button
                            class="btn btn-danger btn-sm"
                            type="button"
                          >
                            Leave only me
                          </button>
                        </div>
                      </div>
                      <div
                        class="mt-3 border border-3 shadow rounded p-3 bg-light"
                      >
                        <h3>
                          Redesign ProductPage to allow more space for visualization of products
                           
                        </h3>
                        <div
                          class="mb-3"
                        >
                          <span>
                            <span
                              class=""
                            >
                              The ProductPage needs a redesign to allow more space for visualization of products.
                            </span>
                          </span>
                        </div>
                        <div />
                        <div
                          class="task-actions"
                        >
                          <div
                            class="dropdown-menu-header dropdown btn-group"
                            role="group"
                          >
                            <button
                              aria-expanded="false"
                              class="dropdown-toggle btn btn-primary btn-sm"
                              type="button"
                            >
                              Move to other column
                            </button>
                          </div>
                          <button
                            class="btn btn-danger btn-sm"
                            type="button"
                          >
                            Unassign Me
                          </button>
                          <button
                            class="btn btn-danger btn-sm"
                            type="button"
                          >
                            Leave only me
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="row g-0 vh-100"
      style="background-image: url(geometric-triangle-shapes-background.jpg); background-size: cover;"
    >
      <div
        class="col-12"
      >
        <div
          class="row g-0 bg-light p-2 shadow align-items-center"
        >
          <div
            class="col-2 col-md-1"
          >
            <a
              class="btn btn-light bi bi-arrow-left border"
              href="/"
            />
          </div>
          <div
            class="col-10 col-md-9 col-lg-10"
          >
            <h3
              class="m-0 text-center"
            >
              Tasks assigned to 
              John Smith
            </h3>
            <p
              class="text-secondary text-center"
            >
              Click on any item to expand/collapse its content
            </p>
          </div>
          <div
            class="col-1 col-md-2 col-lg-1 d-none d-md-block text-center"
          >
            <img
              alt="user"
              class="rounded-circle m-1"
              src="image_1"
              width="80"
            />
          </div>
        </div>
        <div
          class="row g-0 mt-3 justify-content-center overflow-auto tasks"
        >
          <div
            class="col-10"
          >
            <p
              class="bg-white rounded text-center p-1 undefined"
            >
              Completed 
              0
               task
              s
               of 
              2
               (
              0
              %)
            </p>
            <div
              class="accordion"
            >
              <div
                class="accordion-item"
              >
                <h2
                  class="accordion-header"
                >
                  <button
                    aria-expanded="false"
                    class="accordion-button collapsed"
                    type="button"
                  >
                    <h3>
                      Backlog
                      <span
                        class="badge bg-primary ms-2"
                      >
                        2
                      </span>
                    </h3>
                    <p>
                      Prioritized list of features that should be implemented as part of a project.
                    </p>
                  </button>
                </h2>
                <div
                  class="accordion-collapse collapse"
                >
                  <div
                    class="accordion-body"
                  >
                    <div
                      class="mt-3 border border-3 shadow rounded p-3 bg-light"
                    >
                      <h3>
                        Implement ConfirmButton component for ProductModal
                         
                      </h3>
                      <div
                        class="mb-3"
                      >
                        <span>
                          <span
                            class=""
                          >
                            A ConfirmButton component should be implemented to allow confirming selected product features in ProductModal.
                          </span>
                        </span>
                      </div>
                      <p
                        class="text-capitalize"
                      >
                        Priority: 
                        medium
                      </p>
                      <div />
                      <div
                        class="task-actions"
                      >
                        <div
                          class="dropdown-menu-header dropdown btn-group"
                          role="group"
                        >
                          <button
                            aria-expanded="false"
                            class="dropdown-toggle btn btn-primary btn-sm"
                            type="button"
                          >
                            Move to other column
                          </button>
                        </div>
                        <button
                          class="btn btn-danger btn-sm"
                          type="button"
                        >
                          Unassign Me
                        </button>
                        <button
                          class="btn btn-danger btn-sm"
                          type="button"
                        >
                          Leave only me
                        </button>
                      </div>
                    </div>
                    <div
                      class="mt-3 border border-3 shadow rounded p-3 bg-light"
                    >
                      <h3>
                        Redesign ProductPage to allow more space for visualization of products
                         
                      </h3>
                      <div
                        class="mb-3"
                      >
                        <span>
                          <span
                            class=""
                          >
                            The ProductPage needs a redesign to allow more space for visualization of products.
                          </span>
                        </span>
                      </div>
                      <div />
                      <div
                        class="task-actions"
                      >
                        <div
                          class="dropdown-menu-header dropdown btn-group"
                          role="group"
                        >
                          <button
                            aria-expanded="false"
                            class="dropdown-toggle btn btn-primary btn-sm"
                            type="button"
                          >
                            Move to other column
                          </button>
                        </div>
                        <button
                          class="btn btn-danger btn-sm"
                          type="button"
                        >
                          Unassign Me
                        </button>
                        <button
                          class="btn btn-danger btn-sm"
                          type="button"
                        >
                          Leave only me
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
